.github/workflows/test.yml

```
@@ -0,0 +1,64 @@
+name: Test
+
+on:
+  pull_request: {}
+  push:
+    branches: [main]
+  workflow_dispatch:
+    inputs: {}
+
+jobs:
+  tests:
+    strategy:
+      matrix:
+        include:
+        - { python-version: "3.7", os: "ubuntu-latest" }
+        - { python-version: "3.8", os: "ubuntu-latest" }
+        - { python-version: "3.9", os: "ubuntu-latest" }
+      fail-fast: false
+
+    name: Pytest on py${{ matrix.python-version }} (${{ matrix.os }})
+    runs-on: ${{ matrix.os }}
+    env:
+      PYTEST_MARKERS: not concurrent
+
+    steps:
+    - name: Check out the repository
+      uses: actions/checkout@v3.0.2
+
+    - name: Install Poetry
+      env:
+        PIP_CONSTRAINT: .github/workflows/constraints.txt
+      run: |
+        pipx install poetry
+        poetry --version
+
+    - name: Setup Python ${{ matrix.python-version }}
+      uses: actions/setup-python@v3.1.2
+      with:
+        python-version: ${{ matrix.python-version }}
+        architecture: x64
+        cache: 'poetry'
+
+    - name: Upgrade pip
+      env:
+        PIP_CONSTRAINT: .github/workflows/constraints.txt
+      run: |
+        pip install pip
+        pip --version
+
+    - name: Install Dependencies
+      run: |
+        poetry env use "${{ matrix.python-version }}"
+        poetry install
+
+    - name: Run pytest
+      run: |
+        poetry run coverage run --parallel -m pytest -m "$PYTEST_MARKERS"
+
+    - name: Upload coverage data
+      if: always() && (matrix.python-version == '3.9')
+      uses: actions/upload-artifact@v3.1.0
+      with:
+        name: coverage-data-"${{ matrix.python-version }}"
```

@aaronsteers I missed in my first pass 🤦 . `name` acts as a directory for the uploaded artifact, so the actual file name will be unique and generated based on the job name, OS, etc.

```suggestion
        name: coverage-data
```